### PR TITLE
Introduce KUBELET_CRIRUNTIME_ARGS to kubeadm drop-in

### PR DESCRIPTION
**Pull request by sysrich**
**What this PR does / why we need it**:
This PR introduces a KUBELET_CRIRUNTIME_ARGS parameter to be provided by a new (optional) /etc/kubernetes/runtime.conf configuration file.

The intention of the file is to give users, but also distributions, an easy place for defining the CRI runtime to be used by kubernetes

Instead of the current documentation, which advocates manually modifying KUBELET_EXTRA_ARGS, this approach will mean distributions (like openSUSE Kubic) can drop in a configuration file for cri-o, docker, or other runtime of their choice, potentially also making it a user choice from distribution packages.

Currently to be able to give such choice it often requires the installation of multiple runtimes alongside each other, to ensure that at least the 'default' is present..which is wasteful for users at scale who don't want to be filling up their hard disks with redundant CRI runtime binaries.

https://kubernetes.io/docs/reference/setup-tools/kubeadm/kubeadm-init/

The runtime.conf file is not included by the kubernetes rpm/deb packages because the expectation it would be provided by the package providing the runtime, or a supplementary configuration package (eg. kubernetes-kubeadm-crio in openSUSE)

**Special notes for your reviewer**: This is my first PR, please be gentle :)

**Release note**:
```release-note
kubeadm now reads a `KUBELET_CRIRUNTIME_ARGS` parameter from `/etc/kubernetes/runtime.conf`. Users or distribution packagers can use this file to define which CRI runtime is used by kubeadm by populating the parameter with values such as `--container-runtime=remote --container-runtime-endpoint=$RUNTIME_ENDPOINT`. Distributions should package `/etc/kubernetes/runtime.conf` with or alongside their chosen CRI runtime(s), containing appropriate parameters.
```

### CODE REVIEW

**Comment by sysrich** [comment]
Hey @k8s-ci-robot , I work for SUSE, and we signed the CLA, but my account wasn't associated properly - go check again please :)

**Comment by sysrich** [comment]
/area kubeadm
/kind enhancement
@kubernetes/sig-cluster-lifecycle-pr-reviews

**Comment by neolit123** [comment]
/cc @kad 
/cc @bart0sh 
/cc @fabriziopandini 

thanks for the PR
@sysrich

File: build/debs/10-kubeadm.conf [hunk 1]
```diff
@@ -7,5 +7,7 @@ EnvironmentFile=-/var/lib/kubelet/kubeadm-flags.env
 # This is a file that the user can use for overrides of the kubelet args as a last resort. Preferably, the user should use
 # the .NodeRegistration.KubeletExtraArgs object in the configuration files instead. KUBELET_EXTRA_ARGS should be sourced from this file.
 EnvironmentFile=-/etc/default/kubelet
+# This is the file that the user can use to choose an alternative CRI runtime. KUBELET_CRIRUNTIME_ARGS should be sourced from this file.
+EnvironmentFile=-/etc/kubernetes/runtime.conf
```
[the remark this answers is not in the document]
Comment (kad) [reply build/debs/10-kubeadm.conf]: I would put more variables/arguments into file above. No need to create one more file.

[the remark this answers is not in the document]
Comment (neolit123) [reply build/debs/10-kubeadm.conf]: part of @sysrich 's suggestion seems to be:

> distributions (like openSUSE Kubic) can drop in a configuration file for cri-o, docker, or other runtime of their choice, potentially also making it a user choice from distribution packages.

@sysrich did you consider distributing (somehow) custom flags in `/etc/kubernetes/runtime.conf` instead?

[the remark this answers is not in the document]
Comment (kad) [reply build/debs/10-kubeadm.conf]: The question is Kubic is going to use directly those k8s packages as-is or make their own ?

**Comment by fabriziopandini** [comment]
@sysrich Thank for your PR!
I have some background questions about the goal of your PR.

Since you are modifying kubeadm, I assume that 
- you will use kubeadm for creating your cluster/joining nodes
- you already know that kubeadm allows customisation of kubelet CRI via the kubeadm config file (that is translated to the `kubeadm-flags.env` passed to kubelet, see relevant code [here](https://github.com/kubernetes/kubernetes/blob/2744f264f5057be090406a782319697c9fb5a1b6/cmd/kubeadm/app/phases/kubelet/flags.go#L74)).

Could you kindly explain to ma what is the goal of your PR/what your PR add to the existing kubeadm behaviour?

File: build/debs/10-kubeadm.conf (hunk 1 above)
[the remark this answers is not in the document]
Comment (sysrich) [reply build/debs/10-kubeadm.conf]: > @sysrich did you consider distributing (somehow) custom flags in /etc/default/kubelet instead?

@neolit123 @kad I did consider that option, but it is not viable. Consider the following scenario - Kubic contains packages for both CRI-O and Docker. 

We want a situation where if kubeadm is installed AND if CRI-O is installed then CRI-O is configured OR if CRI-O is not installed and Docker is installed, then Docker is configured.

This is best handled by having packages dependencies - only your package manager is going to easily know which runtime is installed plus whether or not kubeadm is installed (and therefore to include the config file which configures the runtime in kubeadm)

Therefore I strongly feel this config file SHOULD be separate from either `/etc/default/kubelet` (aka /etc/sysconfig/kubelet on openSUSE) or the kubeadm config file.

Those should be kept for users to set exactly what they want, the way they want, and how they want - packages should generally avoid touching user configuration files like the plague

> The question is Kubic is going to use directly those k8s packages as-is or make their own ?

Kubic, like all openSUSE projects, wants to keep as close to upstream as possible. Hence this PR. I already have packages carrying this patch that have been tested and worked, but I don't want to merge them into the main openSUSE codebase without consultation with the rest of you, like we're doing now :)

After all, if you persuade me to do it differently (a strong possibility) then I'll just have to rework everything anyway.. :)

**Comment by sysrich** [comment]
@fabriziopandini 

> Thank you for your first k8s contribution

Thanks!

> you will use kubeadm for creating your cluster/joining nodes

That's the plan, yes. More context here: https://kubic.opensuse.org/blog/2018-08-09-newdirection/

Generally speaking I want to keep any openSUSE-specific customisation of kubeadm to a minimum. 

But we do have a pretty strong requirement to be able to support both CRI-O and Docker, and yet we want kubeadm to be configured to use whichever one the user has installed, with minimal user interaction required.

This is the first step in that direction (the next will be figuring out a nice way of `kubectl init` `join` and `reset` automatically knowing what to do about `--cri-socket` when Docker is not being used)

> you already considered using customisation of kubelet CRI via the kubeadm config file (that is translated to the kubeadm-flags.env passed to kubelet, see relevant code here).

> Could you kindly help me in understanding what doses work in the existing kubeadm behaviour (for CRI customization)?

[I think my comment to kad and neolit123 covers my answer to these questions](https://github.com/kubernetes/kubernetes/pull/67391#discussion_r210679753), so I won't repeat it here

But if I need to explain further, I'm more than happy to do so. I'm happy to keep talking here, or if things aren't cleared up by next Tues/Wed I intend to be in both the sig meeting and office hours.

**Comment by fabriziopandini** [comment]
@sysrich I'm really happy to read the kubic news! 
Looking forward for the next sig-meeting!

**Comment by kad** [comment]
@sysrich I think direction of this patch is not completely right. 
As @fabriziopandini mentioned, we have code in kubeadm which will put correct CRI options for kubelet into `kubeadm-flags.env`. Having two places where those options are defined (one by kubeadm code, one by one more file that comes from another package) is a path to the conflicting values.
I'd suggest that we better thing about way of reliably detect, do we have any CRI implementation running and available, and if not, fallback/check that Docker is available. If neither detected and validated -> error.

File: build/debs/10-kubeadm.conf (hunk 1 above)
[the remark this answers is not in the document]
Comment (dixudx) [reply build/debs/10-kubeadm.conf]: Actually we could use current `/etc/default/kubelet` (debian) or `/etc/sysconfig/kubelet` (rpm). No need to have another new file.

**Review by dixudx — CHANGES_REQUESTED** [review]

**Comment by sysrich** [comment]
Closing - immediate needs for openSUSE Kubic will be addressed by playing with /etc/sysconfig/kublet as suggested

But long term I'd like us to look at a more structured model for handling distro-provided config, user-provided config, and user-provided runtime switches. 

After the discussions on this topic in today's SIG-cluster-lifecycle meeting you can expect a long wordy Issue from me as soon as I get all my ideas down in some coherent prose.

Thanks all for making my first PR an enjoyable one :)

**Pull request closed without merging**